ako-gateway-api/k8s/validator.go

```diff
@@ -318,3 +322,119 @@ func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
 	}
 	return true
 }
+
+func ValidateGatewayListenerWithSecret(namespace, name string, deleteFlag bool) {
+	secretNSName := namespace + "/" + name
+	present, gwList := akogatewayapiobjects.GatewayApiLister().GetSecretToGateway(secretNSName)
+
+	if present {
+		clonedGWList := make([]string, len(gwList))
+		copy(clonedGWList, gwList)
+		for _, gwNSName := range gwList {
+			gwNamespace, _, gwName := lib.ExtractTypeNameNamespace(gwNSName)
+			// Check gateway present or not
+			gatewayObj, err := akogatewayapilib.AKOControlConfig().GatewayApiInformers().GatewayInformer.Lister().Gateways(gwNamespace).Get(gwName)
+			if err != nil {
+				// ignore gateway
+				continue
+			}
+			gwStatus := akogatewayapiobjects.GatewayApiLister().GetGatewayToGatewayStatusMapping(gwNamespace + "/" + gwName)
+
+			// go through listeners, check cert is present there or not.
+			validListenerCount := 0
+			certFound := false
+			for listenerIndex, listener := range gatewayObj.Spec.Listeners {
+				if listener.TLS != nil {
+					if (listener.TLS.Mode != nil && *listener.TLS.Mode != gatewayv1.TLSModeTerminate) || len(listener.TLS.CertificateRefs) == 0 {
+						continue
+					}
+					for _, certRef := range listener.TLS.CertificateRefs {
+						// TODO: In future, ref grant impact needs to be considered.
+						certNamespace := namespace
+						if certRef.Namespace != nil {
+							certNamespace = string(*certRef.Namespace)
+						}
+						// add condition for checking gateway status
+						if gwStatus != nil && string(certRef.Name) == name && certNamespace == namespace {
+							setConditions(gwStatus, listenerIndex, gatewayObj.ObjectMeta.Generation, deleteFlag)
+							certFound = true
+							break
+						}
+					}
+				}
+				if gwStatus != nil && !akogatewayapilib.IsListenerInvalid(gwStatus, listenerIndex) {
+					validListenerCount += 1
+				}
+			} // listener loop ends
+			// if cert not found, that means it is stale entry in secret to gateway map
+			if certFound {
+				setGatewayCondition(gwStatus, validListenerCount, len(gatewayObj.Spec.Listeners), gatewayObj.ObjectMeta.Generation)
+			} else {
+				//remove the entry from the clone
+				clonedGWList = utils.Remove(clonedGWList, gwNSName)
+			}
+		} // gw loop ends
+		// update the mapping
+		akogatewayapiobjects.GatewayApiLister().UpdateSecretToGateway(secretNSName, clonedGWList)
+	}
+}
+
+// Mapping has to be taken care between secret and gateway
+func setConditions(gwStatus *gatewayv1.GatewayStatus, index int, generation int64, isDelete bool) {
```

Comment:
Should this function be renamed to setListenerConditions ?

Reply:
done